mavlink_messages: Subscribe to all distance_sensor instances

mortenfyhn opened the pull request:
Continuation/rebase of #14338.

**Describe problem solved by this pull request**
If I connect two Garmin LidarLite sensors to my vehicle, which is supported now with v1.11.0, then I only get data from one of them sent over mavlink. I want both, so I can ask QGroundControl to display the distance data for both (in my case upwards and downwards).

**Describe your solution**
Modify the distance sensor mavlink streamer class to stream from all uorb instances.

**Describe possible alternatives**
None considered.

**Test data / coverage**
Tested by running a Pixhawk 4 with two LidarLites before and after this change. Can see that it works, so I get data from both sensors in QGroundControl.

Before and after:
![before2](https://user-images.githubusercontent.com/6639836/92396227-f7661300-f124-11ea-9302-cec87869b59e.png) ![after2](https://user-images.githubusercontent.com/6639836/92396239-fd5bf400-f124-11ea-9cac-a532ed41c0b5.png)

---

I haven't considered the streaming rate issue discussed previously in #14338 (yet). With this method, and the distance sensor mavlink stream configured to 10 Hz, I get messages at 20 Hz (I guess 10 Hz each). If this is the expected behaviour, then I suppose all is fine. If they should be throttled to 10 Hz total, then we might need some adjustments.

dagar commented:
Relatively good timing, we actually have a slightly better mechanism for handling an array of subscriptions now. https://github.com/PX4/Firmware/pull/15632

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4515,11 +4515,21 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 
 private:
-	uORB::Subscription _distance_sensor_sub{ORB_ID(distance_sensor)};
+	uORB::Subscription _distance_sensor_sub[ORB_MULTI_MAX_INSTANCES] {
```
```suggestion
	uORB::SubscriptionMultiArray<distance_sensor_s> _distance_sensor_subs{ORB_ID::distance_sensor};
```

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4531,44 +4541,48 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 
-		if (_distance_sensor_sub.update(&dist_sensor)) {
-			mavlink_distance_sensor_t msg{};
+		for (int i = 0; i < ORB_MULTI_MAX_INSTANCES; i++) {
```
```suggestion
		for (auto &dist_sens_sub : _distance_sensor_subs) {
```

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4531,44 +4541,48 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
+			distance_sensor_s dist_sensor;
 
-			msg.time_boot_ms = dist_sensor.timestamp / 1000; /* us to ms */
+			if (_distance_sensor_sub[i].update(&dist_sensor)) {
```
```suggestion
			if (dist_sens_sub.update(&dist_sensor)) {
```

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4515,11 +4515,21 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 	unsigned get_size() override
 	{
-		return _distance_sensor_sub.advertised() ? (MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES) : 0;
+		unsigned total_size = 0;
```
In `uORB::SubscriptionMultiArray` there's a method to get the number of advertised.

``` C++
    return size_per_instance * _distance_sensor_subs.advertised_count();
```

mortenfyhn replied:
I'll fix this.

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4531,44 +4541,48 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 
-			mavlink_msg_distance_sensor_send_struct(_mavlink->get_channel(), &msg);
+				msg.current_distance = dist_sensor.current_distance * 1e2f; // m to cm
+				msg.id               = dist_sensor.id;
```
For the sake of mavlink I'm wondering if this id should be the instance?

dagar replied:
I think that's the only thing that makes sense because it's a uint8.

dagar replied:
Actually, in that case you should revert the for loop change above to keep the index, then set `msg.id = i;` here.

mortenfyhn replied:
That might make sense. The ID for LL40LS as an example, is currently from what I can tell computed somehow based on a combination of its configuration, such as bus number and sensor direction. Using the instance number seems reasonable. I'll make the change.

dagar replied:
The actual device id (uint32_t) we use for accel/gyro/mag calibration comes from here. https://github.com/PX4/Firmware/blob/df38e3798c25541d1462d8011fbf0d58904ce4f7/src/lib/drivers/device/Device.hpp#L166-L171

For various reasons like per sensor configuration (orientation, position) I'd like to get that in place for all the distance sensors, but it's still not going to help for mavlink because the id field is only 1 byte.

mortenfyhn commented:
> Relatively good timing, we actually have a slightly better mechanism for handling an array of subscriptions now. #15632

Very nice! I'll update.

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4515,11 +4515,21 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 private:
-	uORB::Subscription _distance_sensor_sub{ORB_ID(distance_sensor)};
+	uORB::SubscriptionMultiArray<distance_sensor_s> _distance_sensor_subs{ORB_ID::distance_sensor};
+		{ORB_ID(distance_sensor), 0}, {ORB_ID(distance_sensor), 1}, {ORB_ID(distance_sensor), 2}, {ORB_ID(distance_sensor), 3}
```
```suggestion
```

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4515,11 +4515,21 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
-	uORB::Subscription _distance_sensor_sub{ORB_ID(distance_sensor)};
+	uORB::SubscriptionMultiArray<distance_sensor_s> _distance_sensor_subs{ORB_ID::distance_sensor};
+		{ORB_ID(distance_sensor), 0}, {ORB_ID(distance_sensor), 1}, {ORB_ID(distance_sensor), 2}, {ORB_ID(distance_sensor), 3}
+	};
```
```suggestion
```

mortenfyhn commented:
Updated now with multi subscriber and ID set to the instance number. Squashed everything and force pushed as one commit.

mortenfyhn commented:
But hang on, regarding the IDs: Do we have control over which physical sensor gets which ID, if we use instance numbers? For mavros, I believe it's necessary to know which sensor will have which ID.

dagar commented:
> But hang on, regarding the IDs: Do we have control over which physical sensor gets which ID, if we use instance numbers? For mavros, I believe it's necessary to know which sensor will have which ID.

We're not even really using the uORB distance_sensor id field and it needs to be replaced with a unique uint32_t device_id like all the other sensors.

The uORB indexing is startup ordering, which will be consistent unless something changes like a sensor being removed or failing to start. If you need more stable ordering than that (long term) we'll need to do something like introduce "calibration" slots of parameters that handle the mapping a specific distance sensor (by device id) to an index.

dagar commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4531,44 +4531,48 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 	bool send(const hrt_abstime t) override
 	{
-		distance_sensor_s dist_sensor;
+		auto updated = false;
```
Nitpicky, but I think the general consensus is to not use `auto` for primitives.

dagar replied:
```suggestion
		bool updated = false;
```

mortenfyhn commented:
> The uORB indexing is startup ordering, which will be consistent unless something changes like a sensor being removed or failing to start.

That should be sufficient.

Fixed the auto -> bool too, and squashed again.
